From 805639235649cfdaf4233a7578545e91259c8aaa Mon Sep 17 00:00:00 2001
From: dianlujitao <dianlujitao@gmail.com>
Date: Sun, 26 Oct 2014 22:37:55 +0800
Subject: [PATCH] init: add more wifi firmware paths

Some devices' wifi firmware is under /data/misc/wifi.
Wlan module won't be loaded successfully if we don't
change this.

A binary "/system/bin/wlan_detect" copies wifi firmware
from /system/etc/wifi to /data/misc/wifi and load it then.

We can't even insmod wlan.ko without this!

Change-Id: Idcbd337c090acb86aefad8c6c69537c8ba8faec6
Signed-off-by: dianlujitao <dianlujitao@gmail.com>
---
 init/devices.c |   31 ++++++++++++++++++++-----------
 1 file changed, 20 insertions(+), 11 deletions(-)

diff --git a/init/devices.c b/init/devices.c
index ce25838..c4953df 100644
--- a/init/devices.c
+++ b/init/devices.c
@@ -51,6 +51,7 @@
 #define FIRMWARE_DIR1   "/etc/firmware"
 #define FIRMWARE_DIR2   "/vendor/firmware"
 #define FIRMWARE_DIR3   "/firmware/image"
+#define FIRMWARE_DIR4   "/data/misc/wifi"
 
 extern struct selabel_handle *sehandle;
 extern char bootdevice[32];
@@ -764,7 +765,7 @@ static int is_booting(void)
 
 static void process_firmware_event(struct uevent *uevent)
 {
-    char *root, *loading, *data, *file1 = NULL, *file2 = NULL, *file3 = NULL;
+    char *root, *loading, *data, *file1 = NULL, *file2 = NULL, *file3 = NULL, *file4 = NULL;
     int l, loading_fd, data_fd, fw_fd;
     int booting = is_booting();
 
@@ -795,6 +796,10 @@ static void process_firmware_event(struct uevent *uevent)
     if (l == -1)
         goto data_free_out;
 
+    l = asprintf(&file4, FIRMWARE_DIR4"/%s", uevent->firmware);
+    if (l == -1)
+        goto data_free_out;
+
     loading_fd = open(loading, O_WRONLY);
     if(loading_fd < 0)
         goto file_free_out;
@@ -810,17 +815,20 @@ try_loading_again:
         if (fw_fd < 0) {
             fw_fd = open(file3, O_RDONLY);
             if (fw_fd < 0) {
-                if (booting) {
-                        /* If we're not fully booted, we may be missing
-                         * filesystems needed for firmware, wait and retry.
-                         */
-                    usleep(100000);
-                    booting = is_booting();
-                    goto try_loading_again;
+                fw_fd = open(file4, O_RDONLY);
+                if (fw_fd < 0) {
+                    if (booting) {
+                            /* If we're not fully booted, we may be missing
+                             * filesystems needed for firmware, wait and retry.
+                             */
+                        usleep(100000);
+                        booting = is_booting();
+                        goto try_loading_again;
+                    }
+                    INFO("firmware: could not open '%s' %d\n", uevent->firmware, errno);
+                    write(loading_fd, "-1", 2);
+                    goto data_close_out;
                 }
-                INFO("firmware: could not open '%s' %d\n", uevent->firmware, errno);
-                write(loading_fd, "-1", 2);
-                goto data_close_out;
             }
         }
     }
@@ -839,6 +847,7 @@ file_free_out:
     free(file1);
     free(file2);
     free(file3);
+    free(file4);
 data_free_out:
     free(data);
 loading_free_out:
-- 
1.7.9.5

